src/file.rs

```diff
@@ -108,111 +110,87 @@ impl TableFile {
-				Ok(0) => break,
-				Ok(n) => {
-					buf = &mut buf[n..];
-					offset += n as u64;
-				},
-				Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {
-					// Try again
-				},
-				Err(e) => return Err(Error::Io(e)),
-			}
-		}
-
-		if !buf.is_empty() {
-			Err(Error::Io(io::Error::new(
-				io::ErrorKind::UnexpectedEof,
-				"failed to fill whole buffer",
-			)))
-		} else {
-			Ok(())
-		}
+	pub fn slice_at(&self, offset: u64, len: usize) -> &[u8] {
+		let offset = offset as usize;
+		let map = self.map.read();
+		let (map, _) = map.as_ref().unwrap();
+		let data: &[u8] = unsafe {
+			let ptr = map.as_ptr().add(offset);
+			std::slice::from_raw_parts(ptr, len)
+		};
+		data
 	}
 
-	#[cfg(windows)]
-	pub fn write_at(&self, mut buf: &[u8], mut offset: u64) -> Result<()> {
-		use crate::error::Error;
-		use std::{io, os::windows::fs::FileExt};
-
-		let file = self.file.read();
-		let file = file.as_ref().unwrap();
-
-		while !buf.is_empty() {
-			match file.seek_write(buf, offset) {
-				Ok(0) =>
-					return Err(Error::Io(io::Error::new(
-						io::ErrorKind::WriteZero,
-						"failed to write whole buffer",
-					))),
-				Ok(n) => {
-					buf = &buf[n..];
-					offset += n as u64;
-				},
-				Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {
-					// Try again
-				},
-				Err(e) => return Err(Error::Io(e)),
-			}
-		}
-
+	pub fn write_at(&self, buf: &[u8], offset: u64) -> Result<()> {
+		let map = self.map.read();
+		let (map, _) = map.as_ref().unwrap();
+		let offset = offset as usize;
+
+		// Nasty mutable pointer cast. We do ensure that all chunks that are being written are
+		// accessed through the overlay in other threads.
+		let ptr: *mut u8 = map.as_ptr() as *mut u8;
+		let data: &mut [u8] = unsafe {
+			let ptr = ptr.add(offset);
+			std::slice::from_raw_parts_mut(ptr, buf.len())
+		};
+		data.copy_from_slice(buf);
 		Ok(())
 	}
 
 	pub fn grow(&self, entry_size: u16) -> Result<()> {
-		let mut capacity = self.capacity.load(Ordering::Relaxed);
-		capacity += GROW_SIZE_BYTES / entry_size as u64;
-
+		let mut map_and_file = self.map.write();
+		let new_len = match map_and_file.as_mut() {
+			None => {
+				let file = self.create_file()?;
+				let len = GROW_SIZE_BYTES;
+				try_io!(file.set_len(len));
+				let mut map = try_io!(unsafe {
+					memmap2::MmapOptions::new().len(RESERVE_ADDRESS_SPACE).map_mut(&file)
+				});
+				madvise_random(&mut map);
+				*map_and_file = Some((map, file));
+				len
+			},
+			Some((map, file)) => {
+				let new_len = try_io!(file.metadata()).len() + GROW_SIZE_BYTES;
+				try_io!(file.set_len(new_len));
+				if map.len() < new_len as usize {
+					let mut new_map = try_io!(unsafe {
+						memmap2::MmapOptions::new().len(map_len(new_len)).map_mut(&*file)
+					});
+					madvise_random(&mut new_map);
+					let old_map = std::mem::replace(map, new_map);
+					try_io!(old_map.flush());
+					// Leak the old mapping as there might be concurrent readers.
+					std::mem::forget(old_map);
```

Comment:
Why is this leak not an issue? I'm trying to identify the root cause of https://github.com/paritytech/parity-db/issues/233 and this looks like a candidate.